src/ui/projectmanager/db/pm.go

```diff
@@ -106,29 +109,107 @@ func (p *ProjectManager) GetRoles(username string, projectIDOrName interface{})
 }
 
 // GetPublic returns all public projects
-func (p *ProjectManager) GetPublic() []models.Project {
-	projects, err := dao.GetProjects("", 1)
-	if err != nil {
-		log.Errorf("failed to get all public projects: %v", err)
-		return []models.Project{}
+func (p *ProjectManager) GetPublic() []*models.Project {
+	return filter("", "", "true", "", 0, 0, 0)
+}
+
+// GetByMember returns all projects which the user is a member of
+func (p *ProjectManager) GetByMember(username string) []*models.Project {
+	return filter("", "", "", username, 0, 0, 0)
+}
+
+// Create ...
+func (p *ProjectManager) Create(project *models.Project) (int64, error) {
+	if project == nil {
+		return 0, fmt.Errorf("project is nil")
 	}
 
-	return projects
+	if len(project.Name) == 0 {
+		return 0, fmt.Errorf("project name is nil")
+	}
+
+	if project.OwnerID == 0 {
+		if len(project.OwnerName) == 0 {
+			return 0, fmt.Errorf("owner ID and owner name are both nil")
+		}
+
+		user, err := dao.GetUser(models.User{
+			Username: project.OwnerName,
+		})
+		if err != nil {
+			return 0, err
+		}
+		if user == nil {
+			return 0, fmt.Errorf("can not get owner whose name is %s", project.OwnerName)
+		}
+		project.OwnerID = user.UserID
+	}
+
+	t := time.Now()
+	pro := &models.Project{
+		Name:         project.Name,
+		Public:       project.Public,
+		OwnerID:      project.OwnerID,
+		CreationTime: t,
+		UpdateTime:   t,
+	}
+
+	return dao.AddProject(*pro)
 }
 
-// GetByMember returns all projects which the user is a member of
-func (p *ProjectManager) GetByMember(username string) []models.Project {
-	user, err := dao.GetUser(models.User{
-		Username: username,
-	})
+// Delete ...
+func (p *ProjectManager) Delete(projectIDOrName interface{}) error {
+	id, ok := projectIDOrName.(int64)
+	if !ok {
+		project := p.Get(projectIDOrName)
+		if project == nil {
+			return fmt.Errorf(fmt.Sprintf("project %v not found", projectIDOrName))
+		}
+		id = project.ProjectID
+	}
+
+	return dao.DeleteProject(id)
+}
+
+// Update ...
+func (p *ProjectManager) Update(projectIDOrName interface{},
+	project *models.Project) error {
+	id, ok := projectIDOrName.(int64)
+	if !ok {
+		pro := p.Get(projectIDOrName)
+		if pro == nil {
+			return fmt.Errorf(fmt.Sprintf("project %v not found", projectIDOrName))
+		}
+		id = pro.ProjectID
+	}
+	return dao.ToggleProjectPublicity(id, project.Public)
+}
+
+// GetAll ...
```

Comment:
need to explain each parm or point to the interface file so user can see the doc.